### ISSUE TITLE
Remove `obtain_from_generic` from all classes

### ISSUE THREAD

**Issue by karlnapf** [shogun-toolbox/shogun#4574]
We now have `.as<T>` which does the job. Easy entrance task

**Comment by Macr0Nerd** [comment]
Do you just want it deleted or replaced with anything? I can take this on and most likely be done by Monday should my schedule permit.

**Comment by vigsterkr** [comment]
@Macr0Nerd remove the functions and wherever it is being used use the mentioned .as<T> function

**Comment by Macr0Nerd** [comment]
I’ll get on that ASAP!

**Comment by Macr0Nerd** [comment]
@vigsterkr where could I find the .as<T>? I just want to make sure I use it right

**Comment by vigsterkr** [comment]
it's part of SGObject

**Comment by Macr0Nerd** [comment]
@vigsterkr I apologize but I will not be able to help with the issue. I have had some things come up and i don't want to leave the issue where it goes unsolved.

**Comment by karlnapf** [comment]
Thanks for letting us know :D

**Comment by remyabel** [comment]
Hi, I'm working on this. What would be the proper syntax for the meta examples? Neither of these work:

```
CombinedKernel learnt_kernel_combined = mmd.get_kernel().as<CombinedKernel>()
CombinedKernel learnt_kernel_combined = mmd.get_kernel().as_CombinedKernel()
```

The first statement is not valid syntax, the second spits out:

```
translate.TranslationFailure: 'Variable learnt_kernel_combined not initialised'
```

**Comment by karlnapf** [comment]
as_kernel, as_features, etc...

**Comment by karlnapf** [comment]
Not sure you need that in the example above though..... get/put work on the base class

**Comment by remyabel** [comment]
> Not sure you need that in the example above though..... get/put work on the base class

After removing all instances of `obtain_from_generic` there are only a few cases where it's being used. The meta examples is one of them. Are there additional changes I need to make before `as_CombinedKernel` can work as intended?

**Comment by karlnapf** [comment]
You would use as_kernel

But actually not even that as you can use get on the sgobject. What calls do you want to make on the object ?

**Comment by remyabel** [comment]
> But actually not even that as you can use get on the sgobject. What calls do you want to make on the object ?

That seems to have worked, thanks. I apologize for not being clear earlier. I was trying to fix these two files (the meta language files):

```
../examples/meta/src/statistical_testing/linear_time_maximum_mean_discrepancy.sg
../examples/meta/src/statistical_testing/quadratic_time_maximum_mean_discrepancy.sg
```

and your suggestion did the trick:

```
-CombinedKernel learnt_kernel_combined = CombinedKernel:obtain_from_generic(mmd.get_kernel())
+CombinedKernel learnt_kernel_combined = mmd.get_kernel()
```

**Comment by karlnapf** [comment]
That won’t compile I think you need to change the type to Kernel

**Comment by karlnapf** [comment]
Also see #4581 where I am refactoring some stuff related

**Comment by karlnapf** [comment]
you  in fact just exclude the meta example for statistical testing in this pr. I will be done with the other one soon and then you can rebate against it

**Comment by remyabel** [comment]
> That won’t compile I think you need to change the type to Kernel

How so? I made sure it compiled before I created the PR.

> you can in fact just exclude the meta example for statistical testing in this pr. I will be done with the other one soon and then you can rebate against it

I will rebase and fix any conflicts after that, then.

**Comment by karlnapf** [comment]
You cannot implicitly convert CKernel* so CConbinedKernel* but your example effectively does that.
Make build_cpp_meta_examples should fail

**Comment by neroangelo296** [comment]
Hi,
When you said remove " 'obtain_from_generic' from all classes " which classes are you referring to.
Preferably could you provide a link.
Thanks!

**Comment by vinx13** [comment]
@neroangelo296 you can `git grep obtain_from_generic`

**Comment by neroangelo296** [comment]
@vinx13 Thanks

**Comment by neroangelo296** [comment]
Just to confirm, is this the change we're looking for?
```
CMKLClassification* CMKLClassification::obtain_from_generic(CMachine* machine)
{
     //code here
}
```
to
```
CMKLClassification* CMachine* machine
{
     //code here
}
```

**Comment by vinx13** [comment]
@neroangelo296 you will replace  `obtain_from_generic` call with  `.as<CMKLClassification>()`, and finally remove `obtain_from_generic` definition

**Comment by neroangelo296** [comment]
So its
```
CMKLClassification* .as<CKMLClassification>(CMachine* machine)
```

**Comment by vinx13** [comment]
@neroangelo296 yes, `as` is a template method that we have already defined in https://github.com/shogun-toolbox/shogun/blob/9537f595679993bd56a833629cba832c97165c5c/src/shogun/base/SGObject.h#L668

**Comment by neroangelo296** [comment]
I've made a pull request. 
Thanks for putting up with me.
